npm_and_yarn/helpers/lib/npm/vulnerability-auditor.js

```diff
@@ -0,0 +1,207 @@
+/* Vulnerability auditor for npm
+ *
+ * Inputs:
+ *  - path to directory containing a package.json and a package-lock.json
+ *  - array of security advisories to audit for,
+ *      [
+ *        { dependency_name: string, affected_versions: [string, ...] },
+ *        ...
+ *      ]
+ *
+ * Outputs:
+ *  - object indicating whether a fix is available and how to achieve it,
+ *      {
+ *        dependency_name: string,
+ *        current_version: string,
+ *        target_version: string,
+ *        fix_available: boolean | object,
+ *        fix_updates: [
+ *          {
+ *            dependency_name: string,
+ *            current_version: string,
+ *            target_version: string
+ *          },
+ *          ...
+ *        ]
+ *      }
+ */
+
+const Arborist = require('@npmcli/arborist')
+const nock = require('nock')
+const { promisify } = require('util');
+const exec = promisify(require('child_process').exec)
+
+async function findVulnerableDependencies(directory, advisories) {
+  const npmConfig = await loadNpmConfig()
+  const caCerts = loadCACerts(npmConfig)
+
+  const arb = new Arborist({
+    path: directory,
+    auditRegistry: 'http://localhost:9999',
+    ca: caCerts,
+    force: true,
+    dryRun: true,
+  })
+
+  const scope = nock('http://localhost:9999')
+    .persist()
+    .post('/-/npm/v1/security/advisories/bulk')
+    .reply(200, convertAdvisoriesToRegistryBulkFormat(advisories))
```

Comment:
Nice hack 🤠  would be nice if you could just pass advisories to Arborist somehow? 🤔

Reply:
It's not tied to the main arborist export but you could try using the same [method](https://github.com/npm/cli/blob/f2340f816f5a137038945229bf8a0d30160e76e8/workspaces/arborist/lib/audit-report.js#L275) that arborist uses to do this:

```js
> AuditReport = require('@npmcli/arborist/lib/audit-report')
[class AuditReport extends Map]
> AuditReport.auditToBulk
[Function: auditToBulk]
```

The arborist method would use the `id` from the advisory itself.  You'd have to move the `nextAdvisoryId` counter logic somewhere else still (if you still need it, I'm not 100% sure what problem that's solving).

Reply:
> You'd have to move the nextAdvisoryId counter logic somewhere else still (if you still need it, I'm not 100% sure what problem that's solving).

Early on we observed that if we returned multiple vuln version ranges for a single dependency, e.g.,

```json
{
  "node-forge": [
    {"vulnerable_versions":"<0.10.0"},
    {"vulnerable_versions":"<1.0.0"},
    {"vulnerable_versions":"<1.3.0"}
  ]
}
```

Arborist would only pick up the first range (`<0.10.0` here) and ignore the rest. We eventually figured out that internally it was uniquing these objects by an `id` field so we added the counter. Thanks for the callout—we could make this clearer in the code comments.

Reply:
> Early on we observed that if we returned multiple vuln version ranges for a single dependency, e.g.,

That's not how advisories return multiple ranges in the npm advisory database. A single semver string is generated that includes all the versions.  Where are the multiple ranges being sourced from? Is this something you're generating to cover use cases? If that's the case I'd highly recommend trying to get your fixture data to mirror that of the npm advisory endpoint.  Having to generate your own ids here seems like a solution to a problem rooted in patterning your fixtures in a way that doesn't match the real advisory data.

Again I admittedly don't know a lot about the internals here, but having an array of `vulnerable_versions` for a single advisory doesn't match what I recall from the npm advisory endpoints.

Reply:
I went back and reproduced the debugging that led us down this path in the first place, by logging the response from the bulk audit endpoint (`/-/npm/v1/security/advisories/bulk`) in Arborist using a local npmcli checkout (`@v8.5.1`).

```javascript
{
  'node-forge': [
    {
      id: 1070354,
      url: 'https://github.com/advisories/GHSA-2r2c-g63r-vccr',
      title: 'Improper Verification of Cryptographic Signature in `node-forge`',
      severity: 'moderate',
      vulnerable_versions: '<1.3.0',
      cwe: [Array],
      cvss: [Object]
    },
    {
      id: 1070355,
      url: 'https://github.com/advisories/GHSA-x4jg-mjrx-434g',
      title: 'Improper Verification of Cryptographic Signature in node-forge',
      severity: 'high',
      vulnerable_versions: '<1.3.0',
      cwe: [Array],
      cvss: [Object]
    },
    {
      id: 1070356,
      url: 'https://github.com/advisories/GHSA-cfm4-qjh2-4765',
      title: 'Improper Verification of Cryptographic Signature in node-forge',
      severity: 'high',
      vulnerable_versions: '<1.3.0',
      cwe: [Array],
      cvss: [Object]
    },
    {
      id: 1067452,
      url: 'https://github.com/advisories/GHSA-8fr3-hfg3-gpgp',
      title: 'Open Redirect in node-forge',
      severity: 'moderate',
      vulnerable_versions: '<1.0.0',
      cwe: [Array],
      cvss: [Object]
    },
    {
      id: 1067471,
      url: 'https://github.com/advisories/GHSA-5rrq-pxf6-6jx5',
      title: 'Prototype Pollution in node-forge debug API.',
      severity: 'low',
      vulnerable_versions: '<1.0.0',
      cwe: [Array],
      cvss: [Object]
    },
    {
      id: 1067473,
      url: 'https://github.com/advisories/GHSA-gf8q-jrpm-jvxq',
      title: 'URL parsing in node-forge could lead to undesired behavior.',
      severity: 'low',
      vulnerable_versions: '<1.0.0',
      cwe: [],
      cvss: [Object]
    }
  ]
}
```

@wraithgar is this not what you're expecting?

Reply:
Ah yes ok I think I understand.  The question now becomes "do the advisories being passed into `findVulnerableDependencies` have their own id attribute?".  If they do then the `auditToBulk` method should be a drop-in replacement for `convertAdvisoriesToRegistryBulkFormat`. If not, then that's another matter altogether.

Reply:
Currently the advisories that get passed to the script do not have an `id` but I wonder we couldn't fix that easily... I'll investigate. Thanks!

Reply:
If you're sourcing them from the github advisory db and not the npm one that would explain it.  Those ids are added by the npm database and are generally unrelated to the advisory itself (in fact currently those ids change if any data in the advisory itself changes).  Arborist et al have no limits on what that id can be, so you may be able to get away with setting the `ghsaId` as the id of the object.

Reply:
fwiw [this is the code](https://github.com/npm/security-microservice/blob/master/lib/utils/github-gql-adapter.js) that pulls from github to npm for advisories.  I think the decision to not tether ids to the ghsaId was not the best, in hindsight, and if this works here (using the ghsaId as id) it could be the beginning of a conversation w/ the registry team to get that rectified.

Reply:
We recognize and agree that mocking the bulk advisory db response is not ideal and we would like to find a better approach as soon as that's feasible. To hopefully satisfy the question for now, I want to provide some additional explanation for our decision.

First, we're trying to resolve at a single dependency at a time. It looks like the default behavior of `npm audit [fix]` is to report/fix advisories on all dependencies in the tree. By mocking the bulk audit response, we're able to narrow the scope of the report/fix to just the dependency we care about and its effects.

Second, the security advisories passed to the script come from the GitHub advisory db. We want to avoid any risk of the npm db being out of sync with ours, and controlling the bulk audit response does give us that.